dev.c

```diff
@@ -135,20 +134,70 @@ static unsigned len_args(unsigned numargs, struct fuse_arg *args)
 
 static u64 fuse_get_unique(struct fuse_conn *fc)
 {
-	fc->reqctr++;
+	struct fuse_per_cpu_ids *my_ids;
+	u64 uid;
+	int num_alloc;
+
+	int cpu = get_cpu();
+
+	my_ids = per_cpu_ptr(fc->per_cpu_ids, cpu);
+
+	if (unlikely(my_ids->num_free_ids == 0)) {
+		spin_lock(&fc->lock);
+		BUG_ON(fc->num_free_ids == 0);
+		num_alloc = min(fc->num_free_ids, (u32)FUSE_MAX_PER_CPU_IDS / 2);
+		memcpy(my_ids->free_ids, &fc->free_ids[fc->num_free_ids - num_alloc],
+			num_alloc * sizeof(u64));
+		fc->num_free_ids -= num_alloc;
+		spin_unlock(&fc->lock);
+
+		my_ids->num_free_ids = num_alloc;
+	}
+
+	uid = my_ids->free_ids[--my_ids->num_free_ids];
+
+	put_cpu();
+
+	uid += FUSE_MAX_REQUEST_IDS;
```

Comment:
I don't get this. We just retrieved a new uid from a stack of existing uids initialized at L891. Why the + FUSE_MAX_REQUEST_IDS. And also where's the corresponding -FUSE_MAX_REQUEST_IDS when putting back?

Reply:
The upper bits serve as a generation count. This prevents the user space from trying to complete the same request twice. The generation count is incremented each time the id is reused. This is what += is for. There is no corresponding -=.